documentation/adr/0004-trip-modifications-mqtt-topic-structure.md

```diff
@@ -0,0 +1,134 @@
+# 4. Trip Modifications MQTT Topic Structure
+
+Date: 2024-07-04
+
+## Status
+
+Draft
+
+## Context
+
+[With the creation of #ADR0003](./0003-use-mqtt-to-publish-trip-modifications.md),
+we are trying to decide on the structure for _how_ we send and persist these 
+[Trip Modifications](https://gtfs.org/realtime/reference/#message-tripmodifications) between Skate and Transit Data.
+
+We know that we'll need these messages to be available if the Transit Data
+service is restarted.
+
+## Decision
+
+We'll use a
+[topic hierarchy](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/)
+which allows us to manage individual GTFS-RT messages and update them, and does
+not require that the Transit Data service has some way to persist state between
+restarts.
+
+### Topic Hierarchy
+We'll nest all of our Trip Modifications related data under the
+[MQTT "topic level"](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/)
+`trip_modifications`, which will be under our "topic prefix"
+configured in the DevOps repo for each environment.
+
+For Example: (spaces used for alignment, and are not part of the topic)
+```
+skate/prod     /trip_modifications/
+skate/dev      /trip_modifications/
+skate/dev-blue /trip_modifications/
+skate/dev-green/trip_modifications/
```

Comment:
```suggestion
     skate/prod/trip_modifications/
      skate/dev/trip_modifications/
 skate/dev-blue/trip_modifications/
skate/dev-green/trip_modifications/
```

The spaces confused me despite the annotation 😢.

This is not a huge deal at all, but if we want the `/trip_modifications/` to all be aligned, maybe right-aligning the whole thing?

Reply:
What do you think about doing something like this?
````diff
diff --git a/documentation/adr/0004-trip-modifications-mqtt-topic-structure.md b/documentation/adr/0004-trip-modifications-mqtt-topic-structure.md
index a512c6f91e...f249ca111e 100644
--- a/documentation/adr/0004-trip-modifications-mqtt-topic-structure.md
+++ b/documentation/adr/0004-trip-modifications-mqtt-topic-structure.md
@@ -29,13 +29,21 @@
 `trip_modifications`, which will be under our "topic prefix"
 configured in the DevOps repo for each environment.
 
-For Example: (spaces used for alignment, and are not part of the topic)
+For Example:
+
+Our `<prefix>`s would be derived from the environment and are described via IaC
+in the devops repo
 ```
-skate/prod     /trip_modifications/
-skate/dev      /trip_modifications/
-skate/dev-blue /trip_modifications/
-skate/dev-green/trip_modifications/
+skate/prod     
+skate/dev      
+skate/dev-blue 
+skate/dev-green
 ```
+And the final combined topic hierarchy would be formed by prepending _one_ of 
+the `<prefix>`s
+```
+<prefix>/trip_modifications/
+```
 
 
 For each new and unique trip modification, we'll generate a unique ID and use
````

Reply:
Love it

I would add something like:

```diff
+And the final combined topic hierarchy would be formed by prepending _one_ of 
+the `<prefix>`s
+```
+<prefix>/trip_modifications/
+```
+For instance,
+```
+skate/dev-blue/trip_modifications/
+```
```

Reply:
https://github.com/mbta/skate/compare/3d9b684ffe2f00c5c9ee208312fe0b769699ed9e..124dfdc2beb3cc2b22d6279b2fdefc5676cd89d1